### ISSUE TITLE
[Bug] Encapsulating MissingZOAUImport in modules may be trapping valuable information

### ISSUE THREAD

**Issue by ddimatos** [ansible-collections/ibm_zos_core#837]
### Is there an existing issue for this?

- [X] There are no existing issues.

### Are the dependencies a supported version?

- [X] The dependencies are supported.

### IBM Z Open Automation Utilities

v1.2.3

### IBM Enterprise Python

v3.11.x

### IBM z/OS Ansible core Version

v1.5.0

### ansible-version

v2.15.x

### z/OS version

v2.5

### Bug description

For reasons unknown at this time, the modules trap the import error and set it to the instance instead of raising it. For example:

```
try:
    from zoautil_py import datasets
except ImportError:
    datasets = MissingZOAUImport()
``` 
instead of:
```
try:
    from zoautil_py import datasets
except ImportError:
   raise MissingZOAUImport() 
````

It was reported that the traceback only showed:
```
TypeError: Value of unknown type: <class \'TypeError\'>, \'MissingZOAUImport\' object is not callable
```
Which is not clear at all. 

When running the raw Ansible ssh -C command you can see information was kept from propagating:
```
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/u/ibmuser/.local/lib/python3.9/site-packages/zoautil_py/jobs.py", line 13, in <module>
    from zoautil_py.core import call_zoau_library
ImportError: CEE3512S An HFS load of module /u/ibmuser/.local/lib/python3.9/site-packages/zoautil_py/core.abi3.so failed. The system return code was 0000000111; the reason code was EF076015.            
Shared connection to hursvl01.hursley.ibm.com closed.
```

The first issue is to evaluate the modules use of MissingZOAUImport () and determine why we can not raise it instead.

The second consideration is perform some added evaluation in the event of this failure and inspect the environment variables, zoau version, python path, pip install (what ever makes sense that can help the user) when this error occurs. 

 Conversation: https://i<>.slack.com/archives/CSLEJ8VGV/p1688498896899089                                              

### Playbook verbosity output.

```
<hursvl01.hursley.ibm.com> ESTABLISH SSH CONNECTION FOR USER: ibmuser
<hursvl01.hursley.ibm.com> SSH: EXEC ssh -C -o ControlMaster=auto -o ControlPersist=60s -o Port=20222 -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o 'User="ibmuser"' -o ConnectTimeout=10 -o 'ControlPath="/root/.ansible/cp/6b86d9c5de"' hursvl01.hursley.ibm.com '/bin/sh -c '"'"'echo ~ibmuser && sleep 0'"'"''
<hursvl01.hursley.ibm.com> (0, b'/u/ibmuser\n', b'')
<hursvl01.hursley.ibm.com> ESTABLISH SSH CONNECTION FOR USER: ibmuser
<hursvl01.hursley.ibm.com> SSH: EXEC ssh -C -o ControlMaster=auto -o ControlPersist=60s -o Port=20222 -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o 'User="ibmuser"' -o ConnectTimeout=10 -o 'ControlPath="/root/.ansible/cp/6b86d9c5de"' hursvl01.hursley.ibm.com '/bin/sh -c '"'"'( umask 77 && mkdir -p "` echo /u/ibmuser/.ansible/tmp `"&& mkdir "` echo /u/ibmuser/.ansible/tmp/ansible-tmp-1688498155.126281-88730-121252264315524 `" && echo ansible-tmp-1688498155.126281-88730-121252264315524="` echo /u/ibmuser/.ansible/tmp/ansible-tmp-1688498155.126281-88730-121252264315524 `" ) && sleep 0'"'"''
<hursvl01.hursley.ibm.com> (0, b'ansible-tmp-1688498155.126281-88730-121252264315524=/u/ibmuser/.ansible/tmp/ansible-tmp-1688498155.126281-88730-121252264315524\n', b'')
Using module file /root/.ansible/collections/ansible_collections/ibm/ibm_zos_core/plugins/modules/zos_job_query.py
<hursvl01.hursley.ibm.com> PUT /root/.ansible/tmp/ansible-local-88713tepebri8/tmpdfk7jlg9 TO /u/ibmuser/.ansible/tmp/ansible-tmp-1688498155.126281-88730-121252264315524/AnsiballZ_zos_job_query.py
<hursvl01.hursley.ibm.com> SSH: EXEC sftp -b - -C -o ControlMaster=auto -o ControlPersist=60s -o Port=20222 -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o 'User="ibmuser"' -o ConnectTimeout=10 -o 'ControlPath="/root/.ansible/cp/6b86d9c5de"' '[hursvl01.hursley.ibm.com]'
<hursvl01.hursley.ibm.com> (0, b'sftp> put /root/.ansible/tmp/ansible-local-88713tepebri8/tmpdfk7jlg9 /u/ibmuser/.ansible/tmp/ansible-tmp-1688498155.126281-88730-121252264315524/AnsiballZ_zos_job_query.py\n', b'')
<hursvl01.hursley.ibm.com> ESTABLISH SSH CONNECTION FOR USER: ibmuser
<hursvl01.hursley.ibm.com> SSH: EXEC ssh -C -o ControlMaster=auto -o ControlPersist=60s -o Port=20222 -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o 'User="ibmuser"' -o ConnectTimeout=10 -o 'ControlPath="/root/.ansible/cp/6b86d9c5de"' hursvl01.hursley.ibm.com '/bin/sh -c '"'"'chmod u+x /u/ibmuser/.ansible/tmp/ansible-tmp-1688498155.126281-88730-121252264315524/ /u/ibmuser/.ansible/tmp/ansible-tmp-1688498155.126281-88730-121252264315524/AnsiballZ_zos_job_query.py && sleep 0'"'"''
<hursvl01.hursley.ibm.com> (0, b'', b'')
<hursvl01.hursley.ibm.com> ESTABLISH SSH CONNECTION FOR USER: ibmuser
<hursvl01.hursley.ibm.com> SSH: EXEC ssh -C -o ControlMaster=auto -o ControlPersist=60s -o Port=20222 -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o 'User="ibmuser"' -o ConnectTimeout=10 -o 'ControlPath="/root/.ansible/cp/6b86d9c5de"' -tt hursvl01.hursley.ibm.com '/bin/sh -c '"'"'/usr/lpp/IBM/cyp/v3r9/pyz/bin/python3 /u/ibmuser/.ansible/tmp/ansible-tmp-1688498155.126281-88730-121252264315524/AnsiballZ_zos_job_query.py && sleep 0'"'"''
<hursvl01.hursley.ibm.com> (1, b'Traceback (most recent call last):
  File "/tmp/ansible_ibm.ibm_zos_core.zos_job_query_payload_yxu4jt27/ansible_ibm.ibm_zos_core.zos_job_query_payload.zip/ansible_collections/ibm/ibm_zos_core/plugins/modules/zos_job_query.py", line 226, in run_module
    File "/tmp/ansible_ibm.ibm_zos_core.zos_job_query_payload_yxu4jt27/ansible_ibm.ibm_zos_core.zos_job_query_payload.zip/ansible_collections/ibm/ibm_zos_core/plugins/modules/zos_job_query.py", line 304, in query_jobs
      File "/tmp/ansible_ibm.ibm_zos_core.zos_job_query_payload_yxu4jt27/ansible_ibm.ibm_zos_core.zos_job_query_payload.zip/ansible_collections/ibm/ibm_zos_core/plugins/module_utils/job.py", line 154, in job_status
        File "/tmp/ansible_ibm.ibm_zos_core.zos_job_query_payload_yxu4jt27/ansible_ibm.ibm_zos_core.zos_job_query_payload.zip/ansible_collections/ibm/ibm_zos_core/plugins/module_utils/job.py", line 200, in _get_job_status
        TypeError: \'MissingZOAUImport\' object is not callable
        
        During handling of the above exception, another exception occurred:
        
        Traceback (most recent call last):
          File "/u/ibmuser/.ansible/tmp/ansible-tmp-1688498155.126281-88730-121252264315524/AnsiballZ_zos_job_query.py", line 107, in <module>
              _ansiballz_main()\r\n  File "/u/ibmuser/.ansible/tmp/ansible-tmp-1688498155.126281-88730-121252264315524/AnsiballZ_zos_job_query.py", line 99, in _ansiballz_main
                  invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)
                    File "/u/ibmuser/.ansible/tmp/ansible-tmp-1688498155.126281-88730-121252264315524/AnsiballZ_zos_job_query.py", line 47, in invoke_module
                        runpy.run_module(mod_name=\'ansible_collections.ibm.ibm_zos_core.plugins.modules.zos_job_query\', init_globals=dict(_module_fqn=\'ansible_collections.ibm.ibm_zos_core.plugins.modules.zos_job_query\', _modlib_path=modlib_path),
                          File "/usr/lpp/IBM/cyp/v3r9/pyz/lib/python3.9/runpy.py", line 210, in run_module
                              return _run_module_code(code, init_globals, run_name, mod_spec)
                                File "/usr/lpp/IBM/cyp/v3r9/pyz/lib/python3.9/runpy.py", line 97, in _run_module_code
                                    _run_code(code, mod_globals, init_globals,
                                      File "/usr/lpp/IBM/cyp/v3r9/pyz/lib/python3.9/runpy.py", line 87, in _run_code
                                          exec(code, run_globals)
                                            File "/tmp/ansible_ibm.ibm_zos_core.zos_job_query_payload_yxu4jt27/ansible_ibm.ibm_zos_core.zos_job_query_payload.zip/ansible_collections/ibm/ibm_zos_core/plugins/modules/zos_job_query.py", line 367, in <module>
                                              File "/tmp/ansible_ibm.ibm_zos_core.zos_job_query_payload_yxu4jt27/ansible_ibm.ibm_zos_core.zos_job_query_payload.zip/ansible_collections/ibm/ibm_zos_core/plugins/modules/zos_job_query.py", line 363, in main
                                                File "/tmp/ansible_ibm.ibm_zos_core.zos_job_query_payload_yxu4jt27/ansible_ibm.ibm_zos_core.zos_job_query_payload.zip/ansible_collections/ibm/ibm_zos_core/plugins/modules/zos_job_query.py", line 233, in run_module
                                                  File "/tmp/ansible_ibm.ibm_zos_core.zos_job_query_payload_yxu4jt27/ansible_ibm.ibm_zos_core.zos_job_query_payload.zip/ansible/module_utils/basic.py", line 1533, in fail_json
                                                    File "/tmp/ansible_ibm.ibm_zos_core.zos_job_query_payload_yxu4jt27/ansible_ibm.ibm_zos_core.zos_job_query_payload.zip/ansible/module_utils/basic.py", line 1506, in _return_formatted
                                                      File "/tmp/ansible_ibm.ibm_zos_core.zos_job_query_payload_yxu4jt27/ansible_ibm.ibm_zos_core.zos_job_query_payload.zip/ansible/module_utils/common/parameters.py", line 928, in remove_values
                                                        File "/tmp/ansible_ibm.ibm_zos_core.zos_job_query_payload_yxu4jt27/ansible_ibm.ibm_zos_core.zos_job_query_payload.zip/ansible/module_utils/common/parameters.py", line 471, in _remove_values_conditions
                                                        TypeError: Value of unknown type: <class \'TypeError\'>, \'MissingZOAUImport\' object is not callable
                                                        ', b'Shared connection to hursvl01.hursley.ibm.com closed.\r\n')
<hursvl01.hursley.ibm.com> Failed to connect to the host via ssh: Shared connection to hursvl01.hursley.ibm.com closed.
<hursvl01.hursley.ibm.com> ESTABLISH SSH CONNECTION FOR USER: ibmuser
<hursvl01.hursley.ibm.com> SSH: EXEC ssh -C -o ControlMaster=auto -o ControlPersist=60s -o Port=20222 -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o 'User="ibmuser"' -o ConnectTimeout=10 -o 'ControlPath="/root/.ansible/cp/6b86d9c5de"' hursvl01.hursley.ibm.com '/bin/sh -c '"'"'rm -f -r /u/ibmuser/.ansible/tmp/ansible-tmp-1688498155.126281-88730-121252264315524/ > /dev/null 2>&1 && sleep 0'"'"''
<hursvl01.hursley.ibm.com> (0, b'', b'')
```

When invoking Ansible's SSH -C  command:
```
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/u/ibmuser/.local/lib/python3.9/site-packages/zoautil_py/jobs.py", line 13, in <module>
    from zoautil_py.core import call_zoau_library
ImportError: CEE3512S An HFS load of module /u/ibmuser/.local/lib/python3.9/site-packages/zoautil_py/core.abi3.so failed. The system return code was 0000000111; the reason code was EF076015.            
Shared connection to hursvl01.hursley.ibm.com closed.
```

**Comment by ddimatos** [comment]
Relates tracker GEH `zoau/issues/12`

**Comment by rexemin** [comment]
I had a hunch that this issue was probably caused by trying to pass sanity tests, and it looks like it. I changed the way we handle a missing ZOAU import, and got this:
```
Running sanity test "import" on Python 3.10
ERROR: Found 1 import issue(s) on python 3.10 which need to be resolved:
ERROR: plugins/modules/zos_copy.py:837:0: traceback: MissingZOAUImport: ZOAU is not properly configured for Ansible. Unable to: import zoautil_py. Ensure environment variables are properly configured in: Ansible for use with ZOAU.
...
ERROR: Found 5 validate-modules issue(s) which need to be resolved:
ERROR: plugins/modules/zos_copy.py:0:0: import-error: Exception attempting to import module for argument_spec introspection, 'ZOAU is not properly configured for Ansible. Unable to
import zoautil_py. Ensure environment variables are properly configured in
Ansible for use with ZOAU.'
```

Using a callable method gets around this, but sometimes breaks and then hides information.

**Comment by ddimatos** [comment]
> Found 5 validate-modules issue(s) which need to be resolved:

Nice investigative work, glad now we know why --> `Found 5 validate-modules issue(s) which need to be resolved:`

**Comment by rexemin** [comment]
The fix in #1042 only fixes `zos_job_query` and `zos_operator`, the rest of the collection would need to be modified too, as the import handler is used throughout every module and module util. We will roll out changes to the way we handle import errors as we migrate to ZOAU 1.3.0 in 2024.
